File: tests/unit/workspace/workspace.test.ts
import { describe, expect, it, vi, beforeEach, afterEach } from 'vitest';
import { createPane } from '../../../src/lib/workspace/pane';
import { parseLayout, splitPane } from '../../../src/lib/workspace/layout-tree';
import { resizeHere } from '../../../src/lib/workspace/resize';
import {
  closeWorkspaceTab,
  createWorkspace,
  openTab,
  splitFocusedPane,
} from '../../../src/lib/workspace/workspace';
import { bootstrapWorkspace } from '../../../src/lib/workspace/workspace-bootstrap';
import { createDefaultTabRegistry } from '../../../src/lib/workspace/tab-registry';

describe('workspace model', () => {
  beforeEach(() => {
    vi.useFakeTimers();
    vi.setSystemTime(new Date('2026-01-02T03:04:05Z'));
  });

  afterEach(() => vi.useRealTimers());

  it('creates and serializes a restorable workspace', () => {
    const workspace = createWorkspace();
    const raw = JSON.parse(JSON.stringify(workspace.layout)) as unknown;
    expect(parseLayout(raw)).toEqual(workspace.layout);
    expect(parseLayout({ type: 'split', children: [] })).toBeUndefined();
  });

  it('boots into welcome plus main startup tabs', () => {
    const workspace = bootstrapWorkspace();
    expect(workspace.layout?.type).toBe('split');
    if (workspace.layout?.type !== 'split') throw new Error('expected split');
    expect(workspace.layout.direction).toBe('vertical');
    expect(workspace.layout.sizes).toEqual([0.4, 0.6]);
    const right = workspace.layout.children[1];
    if (right.type !== 'pane') throw new Error('expected right pane');
    const group = workspace.tabGroups[right.tabGroupId];
    expect(group.tabIds.map((id) => workspace.tabs[id]?.title)).toEqual([
      'Accounts',
      'Relay Settings',
      'Home',
      'Notifications',
      'Tweet',
    ]);
    expect(workspace.tabs[group.activeTabId!]?.title).toBe('Accounts');
    expect(workspace.tabs[workspace.focusedTabId!]?.title).toBe('Welcome');
  });

  it('supports horizontal and vertical nested splits', () => {
    const workspace = createWorkspace();
    const firstPane = workspace.focusedPaneId;
    const horizontal = splitFocusedPane(workspace, 'horizontal');
    const vertical = splitFocusedPane(horizontal, 'vertical');
    if (!vertical.layout) throw new Error('expected layout');
    expect(vertical.layout.type).toBe('split');
    expect(JSON.stringify(vertical.layout)).toContain(firstPane);
    expect(JSON.stringify(vertical.layout)).toContain('vertical');
  });

  it('opens and closes tabs inside a pane group', () => {
    const workspace = createWorkspace();
    const opened = openTab(
      workspace,
      workspace.focusedPaneId,
      'notifications',
      'Notifications',
    );
    expect(Object.keys(opened.tabs)).toHaveLength(2);
    if (!opened.focusedPaneId || !opened.focusedTabId)
      throw new Error('expected focused tab');
    const closed = closeWorkspaceTab(
      opened,
      opened.focusedPaneId,
      opened.focusedTabId,
    );
    expect(Object.keys(closed.tabs)).toHaveLength(1);
  });

  it('resizes split ratios without changing child identity', () => {
    const pane = createPane('group-a');
    const split = splitPane(pane, pane.id, 'horizontal', createPane('group-b'));
    if (split.type !== 'split') throw new Error('expected split');
    const resized = resizeHere(split, 0, 0.2);
    expect(resized.children).toBe(split.children);
    expect(Math.round(resized.sizes.reduce((a, b) => a + b) * 100)).toBe(100);
  });

  it('creates default runtimes with idempotent cleanup', () => {
    const runtime = createDefaultTabRegistry()
      .require('tweet')
      .createRuntime({ tabId: 'tab-a', config: {} });
    runtime.suspend();
    runtime.resume();
    runtime.close();
    runtime.close();
    expect(runtime.metadata().title).toBe('Tweet');
  });
});

File: crates/lkjstr-domain/tests/workspace_test.rs
use std::collections::BTreeMap;

use lkjstr_domain::{
    LayoutNode, NewPaneIds, NewTabIds, SplitDirection, TabKind, WorkspaceIds, bootstrap_workspace,
    close_workspace_tab, convert_tab, create_workspace, ensure_usable_workspace, focus_tab,
    open_tab, split_focused_pane,
};

#[test]
fn boots_into_welcome_plus_main_startup_tabs() -> Result<(), &'static str> {
    let workspace = bootstrap_workspace();
    let Some(LayoutNode::Split(split)) = workspace.layout else {
        return Err("expected split");
    };
    assert_eq!(split.direction, SplitDirection::Vertical);
    assert_eq!(split.sizes, vec![4000, 6000]);
    let LayoutNode::Pane(right) = &split.children[1] else {
        return Err("expected pane");
    };
    let group = workspace.tab_groups.get(&right.tab_group_id);
    let titles = group.map(|item| {
        item.tab_ids
            .iter()
            .filter_map(|id| workspace.tabs.get(id))
            .map(|tab| tab.title.as_str())
            .collect::<Vec<_>>()
    });
    assert_eq!(
        titles,
        Some(vec![
            "Accounts",
            "Relay Settings",
            "Home",
            "Notifications",
            "Tweet"
        ])
    );
    assert_eq!(
        workspace.focused_tab_id.as_deref(),
        Some("bootstrap-welcome-tab")
    );
    Ok(())
}

#[test]
fn opens_focuses_and_splits_tabs() {
    let workspace = create_workspace(ids("ws"), 10);
    let opened = open_tab(
        workspace,
        None,
        TabKind::Notifications,
        NewTabIds {
            tab_id: "note".into(),
        },
        11,
    );
    assert_eq!(opened.focused_tab_id.as_deref(), Some("note"));
    assert_eq!(opened.tabs.len(), 2);

    let focused = focus_tab(opened, "ws-pane", "ws-tab", 12);
    assert_eq!(focused.focused_tab_id.as_deref(), Some("ws-tab"));

    let split = split_focused_pane(
        focused,
        SplitDirection::Horizontal,
        NewPaneIds {
            pane_id: "split-pane".into(),
            group_id: "split-group".into(),
            tab_id: "split-tab".into(),
            split_id: "split-root".into(),
        },
        13,
    );
    assert_eq!(split.focused_pane_id.as_deref(), Some("split-pane"));
    assert!(matches!(split.layout, Some(LayoutNode::Split(_))));
}

#[test]
fn converts_new_tab_choices_without_replacing_the_tab_id() -> Result<(), &'static str> {
    let workspace = create_workspace(ids("ws"), 10);
    let converted = convert_tab(
        workspace,
        "ws-pane",
        "ws-tab",
        TabKind::Profile,
        BTreeMap::from([("pubkey".to_owned(), "abc".to_owned())]),
        11,
    );
    let tab = converted
        .tabs
        .get("ws-tab")
        .ok_or("missing converted tab")?;

    assert_eq!(converted.focused_tab_id.as_deref(), Some("ws-tab"));
    assert_eq!(tab.kind, TabKind::Profile);
    assert_eq!(tab.title, "Profile");
    assert_eq!(tab.config.get("pubkey").map(String::as_str), Some("abc"));
    assert_eq!(tab.created_at, 10);
    Ok(())
}

#[test]
fn closes_tabs_and_recovers_final_tile() {
    let workspace = open_tab(
        create_workspace(ids("ws"), 10),
        None,
        TabKind::Settings,
        NewTabIds {
            tab_id: "settings".into(),
        },
        11,
    );
    let closed = close_workspace_tab(workspace, "ws-pane", "settings", ids("recovery"), 12);
    assert_eq!(closed.tabs.len(), 1);
    assert_eq!(closed.focused_tab_id.as_deref(), Some("ws-tab"));

    let recovered = close_workspace_tab(closed, "ws-pane", "ws-tab", ids("fallback"), 13);
    assert_eq!(recovered.tabs.len(), 1);
    assert!(
        recovered
            .tabs
            .values()
            .any(|tab| tab.kind == TabKind::Welcome)
    );
}

#[test]
fn unusable_workspace_recovers_with_source_identity() {
    let mut workspace = create_workspace(ids("ws"), 10);
    workspace.layout = None;
    workspace.focused_pane_id = None;
    workspace.focused_tab_id = None;

    let recovered = ensure_usable_workspace(workspace, ids("fallback"), 11);

    assert_eq!(recovered.id, "ws");
    assert!(
        recovered
            .tabs
            .values()
            .any(|tab| tab.kind == TabKind::Welcome)
    );
}

fn ids(prefix: &str) -> WorkspaceIds {
    WorkspaceIds {
        workspace_id: prefix.into(),
        pane_id: format!("{prefix}-pane"),
        group_id: format!("{prefix}-group"),
        tab_id: format!("{prefix}-tab"),
    }
}
